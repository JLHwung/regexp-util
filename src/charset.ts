import { Base } from './base';

export type CharsetInput = Charset | CharsetRawInput;

export type CharsetRawInput =
  | string
  | number
  | [string, string]
  | [number, number];

export type CharsetDataUnit = [number, number];

export const charset = (...inputs: CharsetInput[]) => new Charset(...inputs);

export class Charset extends Base {
  public data: CharsetDataUnit[];

  constructor(...inputs: CharsetInput[]) {
    super();

    this.data = [];
    for (const input of inputs) {
      if (input instanceof Charset) {
        this.data.push(...input.data);
      } else {
        this.data.push(normalize(input));
      }
    }
    this._unique();
  }

  public union(...inputs: CharsetInput[]) {
    return new Charset(this, ...inputs);
  }

  protected _is_empty() {
    return this.data.length === 0;
  }

  protected _to_string() {
    const ranges = this.data.map(
      ([start, end]) =>
        start === end ? unicode(start) : `${unicode(start)}-${unicode(end)}`,
    );
    return `[${ranges.join('')}]`;
  }

  private _unique() {
    this.data.sort(compare);

    const new_data: CharsetDataUnit[] = [];

    let last_data_unit: CharsetDataUnit | null = null;
    for (const data_unit of this.data) {
      if (!last_data_unit || last_data_unit[1] + 1 < data_unit[0]) {
        new_data.push(data_unit);
        last_data_unit = data_unit;
      } else {
        new_data.splice(-1, 1, [last_data_unit[0], data_unit[1]]);
        last_data_unit = new_data[new_data.length - 1];
      }
    }

    this.data = new_data;
  }
}

function unicode(char: number) {
  return `\\u{${char.toString(16)}}`;
}

function char_code(char: string) {
  if (char.length !== 1) {
    const display = `${char.length} (${JSON.stringify(char)})`;
    throw new Error(`Expected length = 1, but received ${display}.`);
  }
  return char.charCodeAt(0);
}

function normalize(raw_input: CharsetRawInput) {
  const [normalized] = [raw_input]
    .map(x => (typeof x !== 'object' ? [x, x] : x))
    .map(x => x.map(u => (typeof u === 'string' ? char_code(u) : u)));
  return normalized as CharsetDataUnit;
}

function compare(a: CharsetDataUnit, b: CharsetDataUnit) {
  return a[0] - b[0];
}
