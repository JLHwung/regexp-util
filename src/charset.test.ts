import { charset } from './charset';

test('create: accept char', () => {
  expect(charset('0').toString()).toEqual('[\\u{30}]');
});

test('create: accept number', () => {
  expect(charset(0).toString()).toEqual('[\\u{0}]');
});

test('create: accept char range', () => {
  expect(charset(['0', '9']).toString()).toEqual('[\\u{30}-\\u{39}]');
});

test('create: accept number range', () => {
  expect(charset([0, 9]).toString()).toEqual('[\\u{0}-\\u{9}]');
});

test('create: accept charset', () => {
  expect(charset(charset(0, 9)).toString()).toEqual('[\\u{0}\\u{9}]');
});

test('create: reject string (length !== 1)', () => {
  expect(() => charset('123')).toThrowErrorMatchingSnapshot();
});

test('create: collapse continous chars', () => {
  expect(charset(1, 2, 3, 4).toString()).toEqual('[\\u{1}-\\u{4}]');
});

test('create: collapse intersecting ranges', () => {
  expect(charset([2, 5], [4, 8]).toString()).toEqual('[\\u{2}-\\u{8}]');
});

test('create: order does not matter', () => {
  expect(charset(1, 5, 3, 7).toString()).toEqual('[\\u{1}\\u{3}\\u{5}\\u{7}]');
});

test('union', () => {
  expect(
    charset(1)
      .union(3)
      .toString(),
  ).toEqual('[\\u{1}\\u{3}]');
});
