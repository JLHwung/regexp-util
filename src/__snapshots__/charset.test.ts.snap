// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create: reject string (length !== 1) 1`] = `"Expected length = 1, but received 3 (\\"123\\")."`;
